packages/hashi/src/cookie.js

```diff
@@ -0,0 +1,222 @@
+const sessionOnly = 'session';
+
+export default class Cookie extends BaseShim {
+  constructor(mediator) {
+    super(mediator);
+    Object.assign(this.events, {
+      NOW: 'now',
+    });
+    this.nameSpace = 'cookie';
+    this.__clearData();
+    this.__nowDiff = 0;
+    this.__setData = this.__setData.bind(this);
+    this.__setNowDiff = this.__setNowDiff.bind(this);
+    this.__getCookies = this.__getCookies.bind(this);
+    this.__setItem = this.__setItem.bind(this);
+    this.on(this.events.STATEUPDATE, this.__setData);
+    this.on(this.events.NOW, this.__setNowDiff);
+  }
+
+  __clearData() {
+    const defaultData = {
+      rootCookies: {},
+      byPath: {},
+    };
+    this.__data = Object.assign({}, defaultData);
+  }
+
+  __removeExpiredCookies(data) {
+    const output = {
+      rootCookies: {},
+      byPath: {},
+    };
+
+    const filterCookies = sourceObj => {
+      const out = {};
+      Object.keys(sourceObj).forEach(key => {
+        if (new Date(sourceObj[key].expires) > this.__now()) {
+          out[key] = sourceObj[key];
+        }
+      });
+      return out;
+    };
+    output.rootCookies = filterCookies(data.rootCookies);
+    Object.keys(data.byPath).forEach(path => {
+      const filtered = filterCookies(data.byPath[path]);
+      if (Object.keys(filtered).length) {
+        // Only bother assigning if any cookies actually remain
+        output.byPath[path] = filtered;
+      }
+    });
+    return output;
+  }
+
+  __setData(data) {
+    if (data && data.rootCookies && data.byPath) {
+      this.__data = this.__removeExpiredCookies(data);
+    } else {
+      this.__clearData();
+    }
+  }
+
+  iframeInitialize() {
+    Object.defineProperty(document, this.nameSpace, {
+      get: this.__getCookies,
+      set: this.__setItem,
+      // By definition, the interfaces we are overwriting are configurable
+      // Because otherwise we couldn't configure them.
+      // Make sure ours is also in case someone else is monkey patching,
+      // to avoid causing errors for their code.
+      configurable: true,
+    });
+  }
+
+  get data() {
+    // Return data that should be persisted across sessions
+    const data = {
+      rootCookies: {},
+      byPath: {},
+    };
+
+    function filterCookies(sourceObj) {
+      const output = {};
+      Object.keys(sourceObj).forEach(key => {
+        if (sourceObj[key].expires !== sessionOnly) {
+          output[key] = sourceObj[key];
+        }
+      });
+      return output;
+    }
+    data.rootCookies = filterCookies(this.__data.rootCookies);
+    Object.keys(this.__data.byPath).forEach(path => {
+      const filtered = filterCookies(this.__data.byPath[path]);
+      if (Object.keys(filtered).length) {
+        data.byPath[path] = filtered;
+      }
+    });
+    return this.__removeExpiredCookies(data);
+  }
+
+  __now() {
+    return new Date(Date.now() + this.__nowDiff);
+  }
+
+  __setNowDiff(nowDiff) {
+    this.__nowDiff = nowDiff;
```

Comment:
Can add a comment, for sure.

This is because generally we have found that client side clocks can be really unreliable. Server clocks can also be unreliable, but this allows us to set all times relative to the Kolibri server, and not at the whims of the client (so this is allowing us to import the Kolibri server time functionality without having to directly import this from Kolibri).